stko/optimizers/macromodel.py

```diff
@@ -0,0 +1,1336 @@
+
+    def _convert_maegz_to_mae(self, run_name):
+        """
+        Convert a ``.maegz`` file to a ``.mae`` file.
+
+        Parameters
+        ----------
+        run_name : :class:`str`
+            The name of the run. The files generated by this run will
+            have this name.
+
+        Returns
+        -------
+        None : :class:`NoneType`
+
+        """
+
+        gz_file = gzip.open(f'{run_name}-out.maegz')
+        with open(f'{run_name}.mae', 'wb') as f:
+            f.write(gz_file.read())
+        gz_file.close()
+
+    def _fix_params(self, mol, com):
+        """
+        Fix bond distances and angles in ``.com`` file.
+
+        For each bond distance, bond angle and torsional angle that
+        does not involve a bond created by
+        :meth:`~.Topology.construct`, a "FX" command is added to the
+        body of the ``.com`` file.
+
+        These lines replace the filler line in the main string.
+
+        Parameters
+        ----------
+        mol : :class:`stk.ConstructedMolecule`
+            The molecule which is to be optimized.
+
+        com : :class:`str`
+            The body of the ``.com`` file which is to have fix commands
+            added.
+
+        Returns
+        -------
+        :class:`str`
+            A string holding the body of the ``.com`` file with
+            instructions to fix the various bond distances and angles
+            as described in the docstring.
+
+        """
+
+        fix_block = ''
+        # Add lines that fix the bond distance.
+        fix_block = self._fix_distances(mol, fix_block)
+        # Add lines that fix the bond angles.
+        fix_block = self._fix_bond_angles(mol, fix_block)
+        # Add lines that fix the torsional angles.
+        fix_block = self._fix_torsional_angles(mol, fix_block)
+
+        return com.replace(
+            '!!!BLOCK_OF_FIXED_PARAMETERS_COMES_HERE!!!\n',
+            fix_block
+        )
+
+
+class MacroModelForceField(MacroModel):
+    """
+    Uses MacroModel force fields to optimize molecules.
+
+    """
+
+    def __init__(
+        self,
+        macromodel_path,
+        output_dir=None,
+        restricted=False,
+        timeout=None,
+        force_field=16,
+        maximum_iterations=2500,
+        minimum_gradient=0.05,
+    ):
+        """
+        Initialize a :class:`MacroModelForceField` object.
+
+        Parameters
+        ----------
+        macromodel_path : :class:`str`
+            The full path of the Schrodinger suite within the user's
+            machine. For example, on a Linux machine this may be
+            something like ``'/opt/schrodinger2017-2'``.
+
+        output_dir : :class:`str`, optional
+            The name of the directory into which files generated during
+            the optimization are written, if ``None`` then
+            :func:`uuid.uuid4` is used.
+
+        restricted : :class:`bool`, optional
+            If ``True`` then an optimization is performed only on the
```

Comment:
This doc string needs to specify what actually happens (i.e. bonds with no building block ids)